packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -102,6 +102,121 @@ const double _inputFormLandscapeHeight = 108.0;
 /// [DatePickerMode.day] mode. It defaults to [DatePickerMode.day], and
 /// must be non-null.
 ///
+/// ### State Restoration
+///
+/// Using this method will not enable state restoration for the date picker.
+/// In order to enable state restoration for a date picker, use
+/// [Navigator.restorablePush] or [Navigator.restorablePushNamed] with
+/// [DatePickerDialog].
+///
+/// For more information about state restoration, see [RestorationManager].
+///
+/// {@tool sample --template=freeform}
+///
+/// This sample demonstrates how to create a restorable Material date picker.
+/// This is accomplished by enabling state restoration by specifying
+/// [MaterialApp.restorationScopeId] and using [Navigator.restorablePush] to
+/// push [DatePickerDialog] when the button is tapped.
+///
+/// {@macro flutter.widgets.RestorationManager}
+///
+/// ```dart imports
+/// import 'package:flutter/material.dart';
+/// ```
+///
+/// ```dart
+/// void main() {
+///   runApp(const MyApp());
+/// }
+///
+/// class MyApp extends StatelessWidget {
+///   const MyApp({Key? key}) : super(key: key);
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return const MaterialApp(
+///       restorationScopeId: 'app',
+///       title: 'Restorable Date Picker Demo',
+///       home: MyHomePage(),
+///     );
+///   }
+/// }
+///
+/// class MyHomePage extends StatefulWidget {
+///   const MyHomePage({Key? key}) : super(key: key);
+///
+///   @override
+///   _MyHomePageState createState() => _MyHomePageState();
+/// }
+///
+/// class _MyHomePageState extends State<MyHomePage> with RestorationMixin {
+///   @override
+///   String get restorationId => 'scaffold_state';
+///
+///   final RestorableDateTime _selectedDate = RestorableDateTime(DateTime.now());
+///   late final RestorableRouteFuture<DateTime> _restorableDatePickerRouteFuture = RestorableRouteFuture<DateTime>(
+///     onComplete: _selectDate,
+///     onPresent: (NavigatorState navigator, Object? arguments) {
+///       return navigator.restorablePush(
+///         _datePickerRoute,
+///         arguments: _selectedDate.value.millisecondsSinceEpoch,
+///       );
+///     },
+///   );
+///
+///   static Route<DateTime> _datePickerRoute(
+///     BuildContext context,
+///     Object? arguments,
+///   ) {
+///     return DialogRoute<DateTime>(
+///       context: context,
+///       builder: (BuildContext context) {
+///         return DatePickerDialog(
+///           restorationId: 'date_picker_dialog',
+///           initialEntryMode: DatePickerEntryMode.calendarOnly,
+///           initialDate: DateTime.fromMillisecondsSinceEpoch(arguments! as int),
+///           firstDate: DateTime(2021, 1, 1),
+///           lastDate: DateTime(2022, 1, 1),
```

Comment:
Will this example start to miss-behave in the future when initialDate > lastDate?

Reply:
Given that the initial date is `DateTime.now()` that could indeed be a problem. Probably better to lock the date down for the example.